File: ios/CaddieAI/Models/WeatherData.swift
//
//  WeatherData.swift
//  CaddieAI
//
//  Real-time weather conditions from Open-Meteo.
//

import Foundation

struct WeatherData: Codable, Sendable {
    var temperatureF: Double
    var windSpeedMph: Double
    var windDirectionDegrees: Double  // meteorological: 0=N, 90=E, 180=S, 270=W
    var weatherCode: Int              // WMO weather code
    var fetchedAt: Date

    // MARK: - Wind Strength Mapping

    /// Maps continuous wind speed (mph) to the app's discrete WindStrength enum
    var windStrength: WindStrength {
        switch windSpeedMph {
        case ..<5:   return .none
        case 5..<12: return .light
        case 12..<20: return .moderate
        default:      return .strong
        }
    }

    // MARK: - Wind Relative to Hole Bearing

    /// Given a hole bearing (degrees, 0-360, tee-to-green), compute the
    /// relative wind direction as seen by the golfer playing the hole.
    ///
    /// Wind meteorological convention: wind direction is where wind COMES FROM.
    /// So wind at 0° means wind blowing FROM the north (toward south).
    ///
    /// If the hole plays north (bearing ~0°) and wind comes from north (0°),
    /// the golfer faces a headwind (into).
    func relativeWindDirection(holeBearingDegrees: Double) -> WindDirection {
        var diff = windDirectionDegrees - holeBearingDegrees
        // Normalize to -180...180
        while diff > 180 { diff -= 360 }
        while diff < -180 { diff += 360 }

        let absDiff = abs(diff)

        if absDiff <= 30 {
            return .into
        } else if absDiff >= 150 {
            return .helping
        } else if diff > 0 {
            // Wind source is clockwise from hole bearing
            return .crossRightToLeft
        } else {
            return .crossLeftToRight
        }
    }

    // MARK: - Display Helpers

    var windDescription: String {
        let cardinal = Self.cardinalDirection(from: windDirectionDegrees)
        return "\(Int(windSpeedMph)) mph from \(cardinal)"
    }

    var temperatureDescription: String {
        "\(Int(temperatureF))°F"
    }

    /// Short weather condition from WMO code
    var conditionDescription: String {
        switch weatherCode {
        case 0:          return "Clear"
        case 1, 2, 3:    return "Partly Cloudy"
        case 45, 48:     return "Foggy"
        case 51, 53, 55: return "Drizzle"
        case 61, 63, 65: return "Rain"
        case 71, 73, 75: return "Snow"
        case 80, 81, 82: return "Showers"
        case 95, 96, 99: return "Thunderstorm"
        default:         return "Overcast"
        }
    }

    /// SF Symbol name for the weather condition
    var conditionSymbol: String {
        switch weatherCode {
        case 0:          return "sun.max.fill"
        case 1, 2, 3:    return "cloud.sun.fill"
        case 45, 48:     return "cloud.fog.fill"
        case 51, 53, 55: return "cloud.drizzle.fill"
        case 61, 63, 65: return "cloud.rain.fill"
        case 71, 73, 75: return "cloud.snow.fill"
        case 80, 81, 82: return "cloud.heavyrain.fill"
        case 95, 96, 99: return "cloud.bolt.rain.fill"
        default:         return "cloud.fill"
        }
    }

    /// Whether weather data is still fresh (< 15 minutes old)
    var isFresh: Bool {
        Date().timeIntervalSince(fetchedAt) < 900
    }

    // MARK: - Cardinal Direction

    static func cardinalDirection(from degrees: Double) -> String {
        let directions = ["N", "NNE", "NE", "ENE", "E", "ESE", "SE", "SSE",
                          "S", "SSW", "SW", "WSW", "W", "WNW", "NW", "NNW"]
        let index = Int((degrees + 11.25).truncatingRemainder(dividingBy: 360) / 22.5)
        return directions[index % 16]
    }
}

File: mobile-flutter/lib/core/weather/weather_data.dart
// WeatherData + RelativeWindDirection — KAN-276 (S6) Flutter port
// of the iOS `WeatherData.swift` and Android `WeatherData.kt`.
//
// **Provider:** the natives use Open-Meteo
// (`https://api.open-meteo.com/v1/forecast`) — free, no API key,
// no auth headers, no signed JWTs. Picked because it works
// identically across platforms with zero plumbing. WeatherKit
// (Apple) was rejected for the migration because it requires a
// signed JWT and would force a platform channel just for iOS.
//
// **Units (request):** the natives ask the API for fahrenheit +
// mph. We do the same so the response shape matches what's
// already on the dashboard side.
//
// **Wind direction convention:** Open-Meteo uses the
// **meteorological** convention — `wind_direction_10m` is the
// direction the wind is blowing **FROM**, in compass degrees,
// 0 = north, 90 = east, 180 = south, 270 = west. This matters
// for the projection math below: a hole that points north
// (`bearing = 0°`) hit by a "wind from 0°" reading is a
// **headwind**, not a tailwind.

class WeatherData {
  const WeatherData({
    required this.temperatureF,
    required this.windSpeedMph,
    required this.windDirectionDegrees,
    required this.weatherCode,
    required this.fetchedAtMs,
    required this.latitude,
    required this.longitude,
  });

  /// Temperature in degrees Fahrenheit.
  final double temperatureF;

  /// Wind speed at 10 m above ground, in mph.
  final double windSpeedMph;

  /// Compass direction the wind is blowing FROM, 0..360, with
  /// 0 = north and increasing clockwise. Meteorological convention
  /// (NOT "where the wind is going").
  final double windDirectionDegrees;

  /// WMO weather code (0 = clear, 1-3 = partly cloudy, 45-48 =
  /// fog, 51-67 = drizzle/rain, 71-77 = snow, 95-99 = thunder).
  /// Stored raw — feature code maps to icons / labels.
  final int weatherCode;

  /// Epoch milliseconds when this reading was fetched. Used by
  /// the cache layer for staleness checks.
  final int fetchedAtMs;

  /// Latitude this reading was fetched for. Used by the cache to
  /// match incoming requests against the cached entry's location
  /// (within the ~1 km tolerance the natives use).
  final double latitude;

  /// Longitude this reading was fetched for.
  final double longitude;

  /// Projects this wind reading onto a hole's tee-to-green bearing
  /// and returns the player-relative wind direction. Direct port
  /// of `WeatherData.relativeWindDirection(holeBearingDegrees:)`
  /// from iOS — the same banding (`±30°` headwind, `±150°+`
  /// tailwind, otherwise cross). Android has two implementations
  /// (an 8-band variant in `AutoDetectService` and a 4-band
  /// variant in `HoleAnalysisViewModel`); we port the iOS 4-band
  /// variant because it's the one the existing UI badges already
  /// use.
  ///
  /// `holeBearingDegrees` is the tee-to-green compass bearing,
  /// 0 = north, increasing clockwise.
  RelativeWindDirection relativeWindDirection(double holeBearingDegrees) {
    var diff = windDirectionDegrees - holeBearingDegrees;
    // Normalize to [-180, 180].
    while (diff > 180) {
      diff -= 360;
    }
    while (diff < -180) {
      diff += 360;
    }

    final absDiff = diff.abs();
    if (absDiff <= 30) return RelativeWindDirection.into;
    if (absDiff >= 150) return RelativeWindDirection.helping;
    // diff > 0 means the wind source is clockwise from the hole
    // bearing, which (since the wind is blowing FROM that source)
    // pushes the ball counter-clockwise — i.e. right-to-left.
    return diff > 0
        ? RelativeWindDirection.crossRightToLeft
        : RelativeWindDirection.crossLeftToRight;
  }

  Map<String, dynamic> toJson() => {
        'temperatureF': temperatureF,
        'windSpeedMph': windSpeedMph,
        'windDirectionDegrees': windDirectionDegrees,
        'weatherCode': weatherCode,
        'fetchedAtMs': fetchedAtMs,
        'latitude': latitude,
        'longitude': longitude,
      };

  factory WeatherData.fromJson(Map<String, dynamic> json) {
    return WeatherData(
      temperatureF: (json['temperatureF'] as num).toDouble(),
      windSpeedMph: (json['windSpeedMph'] as num).toDouble(),
      windDirectionDegrees: (json['windDirectionDegrees'] as num).toDouble(),
      weatherCode: (json['weatherCode'] as num).toInt(),
      fetchedAtMs: (json['fetchedAtMs'] as num).toInt(),
      latitude: (json['latitude'] as num).toDouble(),
      longitude: (json['longitude'] as num).toDouble(),
    );
  }
}

/// 4-band wind direction relative to the hole's tee-to-green
/// bearing. Matches the iOS `WindDirection` enum casings — the
/// shot-recommendation engines (KAN-S7) and the caddie screen
/// badges (KAN-S11) consume this enum directly.
enum RelativeWindDirection {
  /// Wind blowing into the player (head-on). The hole bearing
  /// and the wind direction differ by ≤ 30°.
  into,

  /// Wind blowing with the player (tail wind). Bearing and
  /// direction differ by ≥ 150° (i.e. the wind is coming from
  /// roughly behind the player).
  helping,

  /// Cross wind from the right side of the hole, pushing the
  /// ball left.
  crossRightToLeft,

  /// Cross wind from the left side of the hole, pushing the
  /// ball right.
  crossLeftToRight,
}
